tests/test_validation.py

```diff
@@ -244,3 +243,24 @@ def test_validation_on_write(tmp_path):
 
     with pytest.warns(ValidationWarning):
         model.save(file_path)
+
+
+def test_validate_on_assignment(tmp_path):
+    file_path = tmp_path/"test.asdf"
+
+    model = ValidationModel(validate_on_assignment=False)
+    model.meta.string_attribute = 42  # Bad assignment should cause no warning
+
+    with pytest.warns(ValidationWarning):
+        model.save(file_path)
+
+
+def test_validate_on_assignment_with_environ(tmp_path):
+    file_path = tmp_path/"test.asdf"
+
+    os.environ['VALIDATE_ON_ASSIGNMENT'] = '0'
```

Comment:
Instead of using `os.environ`, try using Pytest's `monkeypatch` fixture, which can set an env var only for this test, while leaving the environment of the user (and other tests which might be running in parallel) unchanged.  So something like:

```python
def test_validate_on_assignment_with_environ(tmp_path, monkeypatch):
    file_path = tmp_path/"test.asdf"

    monkeypatch.setenv("VALIDATE_ON_ASSIGNMENT", False)
```

Reply:
Used `monkeypatch` instead of `os.environ` and expanded the tests to capture unset conditions, as well as for `True` and `False` for the environment variable and parameter for class instantiation.